File: selection_sort/selection_sort.py
# Best O(n^2) time | O(1) space
# Avarage O(n^2) time | O(1) space
# Worst O(n^2) time | O(1) space
def selection_sort(array):
  for i in range(len(array)):
    j = i
    smallest_number_idx = i
    while j < len(array):
      if array[smallest_number_idx] > array[j]:
        smallest_number_idx = j
      j += 1
    swap(smallest_number_idx, i, array)
  return array


def swap(i, j, array):
  array[i], array[j] = array[j], array[i]

print(selection_sort([10, 9, 8, 7]))

File: selection_sort/selection_sort.go
package main

import "fmt"

func selectionSort(array []int) []int {
    for i := 0; i < len(array); i++ {
        smallestNumberIdx := i
        for j := i + 1; j < len(array); j++ {
            if array[smallestNumberIdx] > array[j] {
                smallestNumberIdx = j
            }
        }
        swap(smallestNumberIdx, i, array)
    }
    return array
}

func swap(i, j int, array []int) {
    array[i], array[j] = array[j], array[i]
}

func main() {
    array := []int{10, 9, 8, 7}
    sortedArray := selectionSort(array)
    fmt.Println(sortedArray)
}
